Moon data from US Naval Observatory with Reqwest and Serde_json

robbystk opened the pull request:
Here is a function that fetches data on the moon phase from USNO and returns the raw json.

robbystk committed:
Moon data from US Naval Observatory with Reqwest

bmitch commented on src/time_calculations.rs:
```diff
@@ -79,4 +80,39 @@ pub fn get_current_day_of_year() -> u32 {
+    // this is a really slow test and will break without internet
+    fn test_retrieve_moon_data() {
+        // march 2019 supermoon
+        assert_eq!(get_moon_data(Utc.ymd(2019, 03, 17)), "{
```
Would this test actually make an API call to `https://api.usno.navy.mil`? If so we should try to mock it if possible.

robbystk replied:
It does and that is definitely something to change.  Yes, a mock is a good idea.  Another way around it is to break this function into two separate functions, one which simply fetches the data, and one which actually processes it.  That way we don't have to test the fetcher (because it's just a relay), and can quickly create a request to pass to the processor for testing.  It will speed up the tests and make them more-reliable and not dependent on external resources (like the network).  I will definitely refactor this test before the moon phase feature can land.

bmitch replied:
That's awesome @robbystk! That would be great if you could do it. Thanks for your help!

robbystk replied:
I discovered that mocking in rust is not as straightforward as in other languages, so I went the route of splitting up functionality.  There's now a function that gets the json from USNO with reqwest, and one that parses it with serde_json.  There are unit tests for the parsing function that are nice and fast.  I left an integrated test for the full process that actually hits the USNO's API, but it's ignored.

robbystk replied:
I found the Reqwest_mock crate, which allows mocking with Reqwest.  I'll look into using that.

robbystk replied:
Resolved with 7e923968fc42ac8c0023df0540784c22696cd050

robbystk committed:
Manually implement Deserialize to enable json()

Reqwest includes the `json()` method on a response which uses Serde's
`DeserializeOwned` trait on whatever type you want to deserialize to.
So we make up some structs and an enum to hold the data from the json.
The enum is problematic because the moon phases in the json have spaces
in them, and enum variants are not allowed to contain spaces.  Manually
implementing `Deserialize` for the enum seems to be the most-elegant way
around that.

robbystk committed:
Separate data gathering from processing

robbystk committed:
Move moon phase functions to a separate module

robbystk committed:
Tidy up imports in moon_phase module

robbystk committed:
Tidy up moon phase module with the help of rustfmt

robbystk commented:
I realized there is an easier, more-elegant way to process the json data.  Instead of making custom data structures to hold all the data, we can parse the json with serde_json::from_str(), then just pick out what we need. This will also remove serde as a dependency.  

Therefore, I will close this PR for now, and re-open it when the alternate solution is complete.

robbystk committed:
use serde_json instead of serde

robbystk committed:
Remove unused structures from moon phase module

robbystk committed:
Add test for valid json but invalid moon phase

robbystk committed:
Tidy up moon module

robbystk commented:
I've streamlined the json parsing and removed a lot of extraneous stuff.  This still has a test that hits USNO's API but it's ignored.  A solution to that would be to use the [`reqwest_mock`](https://crates.io/crates/reqwest_mock) crate, but maybe that should be another PR?

robbystk committed:
Use reqwest_mock to test without hitting USNO

robbystk committed:
Reduce repetition in tests by re-using stub

robbystk commented:
The test suite now uses a stub to avoid hitting USNO's API.  There is still an ignored test that does hit it, which you may remove if you wish.

I just realized that there's another way to do this.  The way I did it was to create a `MyClient` struct that simply wraps a client whether it's real or a stub.  It then has a method called `get_usno_json` that uses the encapsulated client to perform the request.  Another way to do it would be make the `get_usno_json` function accept a `reqwest_mock::GenericClient` as an argument, then a real or stub client can be passed in as appropriate.  I think the latter implementation might be more-readable than having to declare and `impl` a mostly-empty struct.

robbystk committed:
Tidy up mocking by using a GenericClient
